lib/shared/s3-packages.js

```diff
@@ -0,0 +1,182 @@
+ *   }
+ *
+ *   s3Packages.getLatestVersionFromBucketXML(body, '>=2.5.0').then((latestVersion) => {
+ *     console.log(`The latest version is: ${latestVersion}`);
+ *   });
+ * });
+ */
+exports.getLatestVersionFromBucketXML = (bucketXML, range = '*') => {
+
+  /* eslint-disable lodash/prefer-lodash-method */
+
+  return xml.parseStringAsync(bucketXML)
+
+  /* eslint-enable lodash/prefer-lodash-method */
+
+    .get('ListBucketResult')
+    .get('Contents')
+    .reduce((accumulator, entry) => {
+      const [ name, version ] = _.split(_.first(entry.Key), '/');
+
+      // The `semver` module refuses to apply ranges to prerelease versions
+      // As a workaround, we drop the prerelease tags, if any, apply the range
+      // on that, and keep using the prerelease tag from then on.
+      // See https://github.com/npm/node-semver#prerelease-tags
+      const strippedVersion = `${semver.major(version)}.${semver.minor(version)}.${semver.patch(version)}`;
+
+      if (name === S3_PACKAGE_NAME && semver.satisfies(strippedVersion, range)) {
+        if (_.isNil(accumulator[version])) {
+          accumulator[version] = 1;
+        } else {
+          accumulator[version] += 1;
+        }
+
+        if (accumulator[version] >= NUMBER_OF_PACKAGES
+            && semver.gt(version, accumulator.latest)) {
+          accumulator.latest = version;
+        }
+      }
+
+      return accumulator;
+    }, {
+
+      // This is a good accumulator default value since
+      // every version is semantically greater than this.
+      latest: '0.0.0'
+
+    })
+    .get('latest');
+};
+
+/**
+ * @summary Query the latest available version from an S3 bucket
+ * @function
+ * @private
+ *
+ * @param {String} bucketUrl - an s3 bucket url
+ * @param {Object} options - options
+ * @param {String} options.currentVersion - current version
+ * @param {String} [options.range] - semver range
+ * @fulfil {String} - latest version
+ * @returns {Promise}
+ *
+ * @example
+ * s3Packages.queryLatestVersion(s3Packages.BUCKET_URL_PRODUCTION, {
+ *   currentVersion: '1.0.0-beta.16',
+ *   range: '<2.0.0'
+ * }).then((latestVersion) => {
+ *   console.log(`The latest version is: ${latestVersion}`);
+ * });
+ */
+exports.queryLatestVersion = (bucketUrl, options) => {
+  return request.getAsync(bucketUrl).then((response) => {
+    return exports.getLatestVersionFromBucketXML(response.body, options.range);
+  }).then((latestVersion) => {
+    if (semver.gte(options.currentVersion, latestVersion)) {
+      return options.currentVersion;
+    }
+
+    return latestVersion;
+  })
+  .catch({
+    code: 'ENOTFOUND'
+  }, {
+    code: 'ETIMEDOUT'
+  }, () => {
+    return options.currentVersion;
+  });
+};
+
+/**
+ * @summary Query the latest available version from an S3 bucket (memoized)
+ * @function
+ * @public
+ *
+ * @description
+ * This function is the memoized version of `.queryLatestVersion()`. The
+ * non-memoized version is kept separately for testability purposes.
```

Comment:
For testing purposes, is there not something we can do to 'clear' the memoization?
That might make the tests messy, but IMHO it'd be less confusing than having both memoized and non-memoized functions which do very similar things.